resource-manager/src/main/java/org/snomed/module/storage/ModuleStorageCoordinator.java

```diff
@@ -0,0 +1,602 @@
+        }
+
+        return rf2Package == null;
+    }
+
+    private String getMetadataResourcePath(String directory, String codeSystem, String moduleId, String effectiveTime) {
+        return getBaseResourcePath(directory, codeSystem, moduleId, effectiveTime) + "/metadata.json";
+    }
+
+    private String getPackageResourcePath(String directory, String codeSystem, String moduleId, String effectiveTime, String rf2PackageFileName) {
+        return getBaseResourcePath(directory, codeSystem, moduleId, effectiveTime) + "/" + rf2PackageFileName;
+    }
+
+    private FileInputStream asFileInputStream(File file) {
+        try {
+            return new FileInputStream(file);
+        } catch (FileNotFoundException e) {
+            LOGGER.trace("Cannot convert File to FileInputStream", e);
+            throw new RuntimeException("Cannot convert File to FileInputStream.", e);
+        }
+    }
+
+    private boolean invalidArguments(String codeSystem, String moduleId, String effectiveTime) {
+        if (codeSystem == null || codeSystem.isBlank()) {
+            return true;
+        }
+
+        if (moduleId == null || moduleId.isBlank()) {
+            return true;
+        }
+
+        return effectiveTime == null || effectiveTime.isEmpty() || !effectiveTime.matches("^\\d{4}\\d{2}\\d{2}$");
+    }
+
+    private String getBaseResourcePath(String directory, String codeSystem, String moduleId, String effectiveTime) {
+        return directory + "/" + codeSystem + "_" + moduleId + "/" + effectiveTime;
+    }
+
+    private ModuleMetadata doBuild(String codeSystem, String moduleId, String effectiveTime, File rf2Package) throws IOException {
+        Set<String> uniqueModuleIds = rf2Service.getUniqueModuleIds(rf2Package);
+        if (uniqueModuleIds.isEmpty()) {
+            throw new RuntimeException("Cannot build ModuleMetadata; no modules found");
+        }
+
+        List<ModuleMetadata> dependencies = new ArrayList<>(getDependencies(rf2Package, uniqueModuleIds));
+
+        ModuleMetadata moduleMetadata = new ModuleMetadata();
+        moduleMetadata.setFilename(rf2Package.getName());
+        moduleMetadata.setCodeSystemShortName(codeSystem);
+        moduleMetadata.setIdentifyingModuleId(moduleId);
+        moduleMetadata.setCompositionModuleIds(new ArrayList<>(uniqueModuleIds));
+        moduleMetadata.setEffectiveTime(asInteger(effectiveTime));
+        moduleMetadata.setFileTimeStamp(new Date(rf2Package.lastModified()));
+        moduleMetadata.setFileMD5(FileUtils.getMD5(rf2Package));
+        moduleMetadata.setPublished(false); // Cannot infer from File alone; subsequent manual updates to metadata required.
+        moduleMetadata.setEdition(false); // Cannot infer from File alone; subsequent manual updates to metadata required.
+        moduleMetadata.setDependencies(dependencies);
+
+        return moduleMetadata;
+    }
+
+    private Set<ModuleMetadata> getDependencies(File rf2Package, Set<String> excludedModuleIds) {
```

Comment:
@pgwilliams This method processes all entries in the MDRS. If an entry in the MDRS is not found in its own package, then it's deemed a dependency. It's presumed all dependencies are uploaded first.